compiler/src/org.graalvm.compiler.truffle.runtime/src/org/graalvm/compiler/truffle/runtime/debug/CompilationStatusListener.java

```diff
@@ -0,0 +1,98 @@
+ *
+ * Please contact Oracle, 500 Oracle Parkway, Redwood Shores, CA 94065 USA
+ * or visit www.oracle.com if you need additional information or have any
+ * questions.
+ */
+package org.graalvm.compiler.truffle.runtime.debug;
+
+import com.oracle.truffle.api.frame.Frame;
+import com.oracle.truffle.api.instrumentation.CompilationState;
+import com.oracle.truffle.api.instrumentation.CompilationStateBackdoor;
+import org.graalvm.compiler.truffle.common.TruffleCompilerListener.CompilationResultInfo;
+import org.graalvm.compiler.truffle.common.TruffleCompilerListener.GraphInfo;
+import org.graalvm.compiler.truffle.runtime.*;
+
+import java.util.concurrent.atomic.AtomicInteger;
+
+public final class CompilationStatusListener extends AbstractGraalTruffleRuntimeListener {
+
+    private final AtomicInteger finished = new AtomicInteger();
+    private final AtomicInteger failures = new AtomicInteger();
+    private final AtomicInteger dequeues = new AtomicInteger();
+    private final AtomicInteger deoptimizations = new AtomicInteger();
+
+    private CompilationStatusListener(GraalTruffleRuntime runtime) {
+        super(runtime);
+        CompilationStateBackdoor.ACCESSOR = this::sampleCompilationState;
+    }
+
+    public static void install(GraalTruffleRuntime runtime) {
+        runtime.addListener(new CompilationStatusListener(runtime));
+    }
+
+    @Override
+    public synchronized void onCompilationDequeued(OptimizedCallTarget target, Object source, CharSequence reason, int tier) {
+        dequeues.getAndIncrement();
+    }
+
+    @Override
+    public synchronized void onCompilationInvalidated(OptimizedCallTarget target, Object source, CharSequence reason) {
+        // We count invalidations as deoptimizations
+        deoptimizations.getAndIncrement();
+    }
+
+    @Override
+    public synchronized void onCompilationSuccess(OptimizedCallTarget target, TruffleInlining inliningDecision, GraphInfo graph, CompilationResultInfo result, int tier) {
+        finished.getAndIncrement();
+    }
+
+    @Override
+    public synchronized void onCompilationFailed(OptimizedCallTarget target, String reason, boolean bailout, boolean permanentBailout, int tier) {
+        if (TraceCompilationListener.isPermanentFailure(bailout, permanentBailout)) {
+            failures.getAndIncrement();
+        }
+    }
+
+    @Override
+    public synchronized void onCompilationDeoptimized(OptimizedCallTarget target, Frame frame) {
+        deoptimizations.getAndIncrement();
+    }
+
+    public synchronized CompilationState sampleCompilationState() {
+        /*
+         * Thermometer TODO: I'd like to track queue size and running count by
+         * using the listener as normal, but I can't find a way to make that
+         * work - I think the key problem is that compilations dequeued may
+         * either be queued or running and I can't tell which. We'll work to
+         * improve that if this is to be merged.
+         */
```

Comment:
If we could change how the compilation listener worked we could use it more simply and not have the `getCompilationsRunning()` API.

Reply:
I suggested a compilation listener API in a different comment.